[stress] Is it possible resources are being deleted out from underneath tests?

I've had a few test runs of azservicebus and azeventhubs where it seems like the Service Bus or Event Hub went down in a much harder way than expected (theory: it's being deleted out from under the test):

This is a log from one of the pods in the playground sublscription:

Info:
* ns: rpeh
* pod: processorinfinite-goeh-4-4pmrt
* pg cluster

```
  [30] o:87724 (last: 2022-10-04T01:57:55Z), remaining: 900/5000
  [31] o:87724 (last: 2022-10-04T01:57:55Z), remaining: 800/5000
2022/10/04 01:58:05 [c(87724), p(26)]: Fatal error from ReceiveEvents: (connlost): received frame header with invalid size 0
2022/10/04 01:58:05 [c(87724), p(26)]: Closing
panic: (connlost): received frame header with invalid size 0
```

I need to track down some more for Service Bus to see, but I think we might want to consider making stress-watcher (or whatever deletes resources) less aggressive so we can narrow this down a bit.

CC: @ckairen, @benbp

--- a/tools/stress-cluster/chaos/README.md
+++ b/tools/stress-cluster/chaos/README.md
@@ -16,6 +16,7 @@ The chaos environment is an AKS cluster (Azure Kubernetes Service) with several 
         * [Customize Docker Build](#customize-docker-build)
      * [Manifest Special Fields](#manifest-special-fields)
      * [Job Manifest](#job-manifest)
+        * [Built-In Labels](#built-in-labels)
      * [Chaos Manifest](#chaos-manifest)
      * [Scenarios and values.yaml](#scenarios-and-valuesyaml)
      * [Node Size Requirements](#node-size-requirements)
@@ -309,6 +310,8 @@ In the example below, the "deploy-job-template.from-pod" template is used, which
 {{- define "stress.deploy-example" -}}
 metadata:
   labels:
+    # Add this label to keep test resources around after test completion until their DeleteAfter tag expiry
+    # Skip.RemoveTestResources: "true"
     testName: "deploy-example"
 spec:
   containers:
@@ -325,6 +328,11 @@ spec:
       {{- include "stress-test-addons.container-env" . | nindent 6 }}
 {{- end -}}
 ```
+
+#### Built-In Labels
+
+- `chaos` - set this to "true" to enable chaos for your pod
+- `Skip.RemoveTestResources` set this to "true" to prevent resources from being deleted immediately after test completion
 
 ### Chaos Manifest
 

--- a/tools/stress-cluster/chaos/examples/stress-deployment-example/templates/deploy-job.yaml
+++ b/tools/stress-cluster/chaos/examples/stress-deployment-example/templates/deploy-job.yaml
@@ -3,6 +3,8 @@
 metadata:
   labels:
     testName: "deploy-example"
+    # Add this label to keep test resources around after test completion until their DeleteAfter tag expiry
+    # Skip.RemoveTestResources: "true"
 spec:
   containers:
     - name: deployment-example

--- a/tools/stress-cluster/services/Stress.Watcher/src/PodEventHandler.cs
+++ b/tools/stress-cluster/services/Stress.Watcher/src/PodEventHandler.cs
@@ -281,18 +281,16 @@ namespace Stress.Watcher
                 return false;
             }
 
-            if (pod.Status.Phase == "Succeeded" || pod.Status.Phase == "Failed")
+            bool isCompleted = (pod.Status.Phase == "Succeeded" || pod.Status.Phase == "Failed");
+            if (isCompleted &&
+                pod.Metadata.Labels.TryGetValue("Skip.RemoveTestResources", out var skipRemove) &&
+                skipRemove == "true")
             {
-                if (pod.Metadata.Labels.TryGetValue("Skip.RemoveTestResources", out var skipRemove) && skipRemove == "true")
-                {
-                    Logger.Information($"Resource has Skip.RemoveTestResources=true label, skipping resource deletion.");
-                    return false;
-                }
-
-                return true;
+                Logger.Information($"Resource has Skip.RemoveTestResources=true label, skipping resource deletion.");
+                return false;
             }
 
-            return false;
+            return isCompleted;
         }
 
         public string GetResourceGroupName(V1Pod pod)

--- a/tools/stress-cluster/services/Stress.Watcher/src/PodEventHandler.cs
+++ b/tools/stress-cluster/services/Stress.Watcher/src/PodEventHandler.cs
@@ -281,7 +281,18 @@ namespace Stress.Watcher
                 return false;
             }
 
-            return (pod.Status.Phase == "Succeeded") || (pod.Status.Phase == "Failed");
+            if (pod.Status.Phase == "Succeeded" || pod.Status.Phase == "Failed")
+            {
+                if (pod.Metadata.Labels.TryGetValue("Skip.RemoveTestResources", out var skipRemove) && skipRemove == "true")
+                {
+                    Logger.Information($"Resource has Skip.RemoveTestResources=true label, skipping resource deletion.");
+                    return false;
+                }
+
+                return true;
+            }
+
+            return false;
         }
 
         public string GetResourceGroupName(V1Pod pod)

--- a/tools/stress-cluster/services/Stress.Watcher/src/Program.cs
+++ b/tools/stress-cluster/services/Stress.Watcher/src/Program.cs
@@ -47,13 +47,13 @@ namespace Stress.Watcher
 
             DotEnv.Load(options: new DotEnvOptions(envFilePaths: new[] {"/mnt/outputs/.env"}));
             var subscriptionId = Environment.GetEnvironmentVariable("AZURE_SUBSCRIPTION_ID");
-            // Default to 'Azure SDK Developer Playground' subscription when testing locally outside of the stress cluster. 
+            // Default to 'Azure SDK Developer Playground' subscription when testing locally outside of the stress cluster.
             subscriptionId = subscriptionId ?? "faa080af-c1d8-40ad-9cce-e1a450ca5b57";
 
             ArmClient armClient = new ArmClient(subscriptionId, new DefaultAzureCredential());
 
             var podEventHandler = new PodEventHandler(client, chaosClient, armClient, options.Namespace);
-            
+
             var cts = new CancellationTokenSource();
             await podEventHandler.Watch(cts.Token);
         }

--- a/tools/stress-cluster/services/Stress.Watcher/tests/PodEventHandlerTest.cs
+++ b/tools/stress-cluster/services/Stress.Watcher/tests/PodEventHandlerTest.cs
@@ -186,6 +186,10 @@ namespace Stress.Watcher.Tests
             handler.ShouldDeleteResources(pod).Should().BeTrue();
             pod.Status.Phase = "Unknown";
             handler.ShouldDeleteResources(pod).Should().BeFalse();
+
+            pod.Status.Phase = "Failed";
+            pod.Metadata.Labels.Add("Skip.RemoveTestResources", "true");
+            handler.ShouldDeleteResources(pod).Should().BeFalse();
         }
 
         [Fact]

